aws_topology/stackstate_checks/aws_topology/resources/sns.py

```diff
@@ -1,58 +1,118 @@
+        data = self.collect_topic_description(topic_arn) or topic_data
+        # If ARN collection fails, skip tag collection
+        if topic_arn:
+            tags = self.collect_tags(topic_arn)
+        else:
+            tags = []
+        return TopicData(topic=data, tags=tags)
+
+    @set_required_access_v2("sns:ListTopics")
+    def collect_topics(self):
+        for topic in [
+            self.collect_topic(topic_data)
+            for topic_data in client_array_operation(self.client, "list_topics", "Topics")
+        ]:
+            yield topic
+
+    def collect_subscriptions(self, topic_arn):
+        for subscription in client_array_operation(
+            self.client, "list_subscriptions_by_topic", "Subscriptions", TopicArn=topic_arn
+        ):
+            yield subscription
 
     def process_all(self, filter=None):
-        for topic_page in self.client.get_paginator('list_topics').paginate():
-            for topic_data_raw in topic_page.get('Topics') or []:
-                topic_data = make_valid_data(topic_data_raw)
-                self.process_topic(topic_data)
-
-    def process_one_topic(self, arn):
-        self.process_topic({"TopicArn":  arn})
-
-    def process_topic(self, topic_data):
-        topic_arn = topic_data['TopicArn']
-        topic_name = topic_arn.rsplit(':', 1)[-1]
-        topic_data['Name'] = topic_name
-        topic_data.update(with_dimensions([{'key': 'TopicName', 'value': topic_name}]))
-        topic_data["Tags"] = self.client.list_tags_for_resource(ResourceArn=topic_arn).get('Tags') or []
-        self.emit_component(topic_arn, self.COMPONENT_TYPE, topic_data)
-        for subscriptions_by_topicpage in self.client.get_paginator('list_subscriptions_by_topic').paginate(
-                TopicArn=topic_arn):
-            for subscription_by_topic in subscriptions_by_topicpage.get('Subscriptions') or []:
-                if subscription_by_topic['Protocol'] in ['lambda', 'sqs'] and \
-                        subscription_by_topic['TopicArn'] == topic_arn:
+        if not filter or "aws.sns" in filter:
+            for topic_data in self.collect_topics():
+                try:
+                    self.process_topic(topic_data)
+                except Exception:
+                    pass
+
+    def process_one_topic(self, topic_arn):
+        self.process_topic(self.collect_topic({"TopicArn": topic_arn}))
+
+    @transformation()
+    def process_topic(self, data):
+        topic = Topic(data.topic, strict=False)
+        topic.validate()
+        output = make_valid_data(data.topic)
+        topic_arn = topic.TopicArn
+        topic_name = topic_arn.rsplit(":", 1)[-1]
+        output["Name"] = topic_name
+        output["Tags"] = data.tags
+        output.update(with_dimensions([{"key": "TopicName", "value": topic_name}]))
+        self.emit_component(topic_arn, self.COMPONENT_TYPE, output)
+
+        for subscription_by_topic in self.collect_subscriptions(topic_arn):
```

Comment:
You should never "collect" inside a "process" that does transformation
collect_subscriptions should be done in collect_topic